### DISCUSSION TITLE
Loop short wav with low overhead

### DISCUSSION THREAD

**Discussion by Benik3** [earlephilhower/ESP8266Audio#612] [Q&A]
Hello.
What is the best way if I want to loop one wav file (in PROGMEM), which is very small (100ms) and should loop almost immediately?
I would like to simulate shots of machine gun (cosplay prop).

Because the shots are so short I would like to do it in the best possible way according to performance.

When I tried to simply do
```
void loop()
{
  if (wav->isRunning()) {
    wav->loop();
  } else {
    Serial.printf("WAV done\n");
    wav->begin(file, out);
  }
}
```

I get error `AudioGeneratorWAV::begin: file not open`. Isn't there more effective way then loading everything again?
Also I wasn't able to find any API documentation.
Thanks.

**Comment by artificiel** [comment]
you have to delete the `file` and make a new one. and you probably have to call `stop()` the wav before re-use, and you may also have to delete and recreate the `wav`, although that does not seem to be always required.

you also have the option of using `seek(0, SEEK_SET)` but superficial testing seems to make a longer interstitial delay than delete/new.

also your output buffer size will have an impact on general latency.

**Reply by dbu** [reply]
seek did not work for me. but instead of delete/new the source, i found that calling close and open on the AudioFileSource before calling begin works. i am no expert, but would assume that close/open causes less memory fragmentation than delete/new on the whole AudioFileSource.

**Reply by Benik3** [reply]
Can be, but still the most of the memory allocation will be probably made when opening the file and then de-allocated when closing.
In my project I ended with Arduino Audio Tools library.
Then I do just:
```
....
if(restart_sound_rising_edge)
{
    play_sound = true;
    music.begin();
}

if (play_sound)
{
    if (!copier.copy())
    {
        play_sound = false;
    }
}
```
